internal/controller/humiogroup_controller.go

```
@@ -0,0 +1,208 @@
+			return reconcile.Result{}, r.logErrorAndReturn(setStateErr, "unable to set cluster state")
+		}
+		return reconcile.Result{RequeueAfter: 5 * time.Second}, r.logErrorAndReturn(err, "unable to obtain humio client config")
+	}
+	humioHttpClient := r.HumioClient.GetHumioHttpClient(cluster.Config(), req)
+
+	// delete
+	r.Log.Info("checking if group is marked to be deleted")
+	isMarkedForDeletion := hg.GetDeletionTimestamp() != nil
+	if isMarkedForDeletion {
+		r.Log.Info("group marked to be deleted")
+		if helpers.ContainsElement(hg.GetFinalizers(), humioFinalizer) {
+			_, err := r.HumioClient.GetGroup(ctx, humioHttpClient, req, hg)
+			if errors.As(err, &humioapi.EntityNotFound{}) {
+				hg.SetFinalizers(helpers.RemoveElement(hg.GetFinalizers(), humioFinalizer))
+				err := r.Update(ctx, hg)
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+				r.Log.Info("Finalizer removed successfully")
+				return reconcile.Result{Requeue: true}, nil
+			}
+
+			// Run finalization logic for humioFinalizer. If the
+			// finalization logic fails, don't remove the finalizer so
+			// that we can retry during the next reconciliation.
+			r.Log.Info("Deleting Group")
+			if err := r.HumioClient.DeleteGroup(ctx, humioHttpClient, req, hg); err != nil {
+				return reconcile.Result{}, r.logErrorAndReturn(err, "Delete group returned error")
+			}
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// Add finalizer for this CR
+	if !helpers.ContainsElement(hg.GetFinalizers(), humioFinalizer) {
+		r.Log.Info("Finalizer not present, adding finalizer to group")
+		hg.SetFinalizers(append(hg.GetFinalizers(), humioFinalizer))
+		err := r.Update(ctx, hg)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+	defer func(ctx context.Context, hg *humiov1alpha1.HumioGroup) {
+		_, err := r.HumioClient.GetGroup(ctx, humioHttpClient, req, hg)
+		if errors.As(err, &humioapi.EntityNotFound{}) {
+			_ = r.setState(ctx, humiov1alpha1.HumioGroupStateNotFound, hg)
```

We're ignoring the errors here? Maybe at least log it?

---

I believe the "finalizer" acts on this error state "eventually". for this part I am mostly copying the existing logic from other controllers, would love if someone from CrowdStrike can confirm this is correct though